rfd/0139-gogo-must-go.md

```diff
@@ -97,15 +97,50 @@ Serializing the objects in the backend as the gRPC wire format has a number of b
 - Allows us to remove gogoproto extensions without affecting backend storage.
 - Allows us to utilize fields in the backend more efficiently and store larger objects.
 
-The primary disadvantage here is that manually examining the backend is more difficult as
-the stored values will no longer be human readable.
+The disadvantages here:
+
+- Manually examining the backend is more difficult as the stored values will no longer be
+  human readable.
+- For support cases we will occasionally need to perform surgery in the backend, which
+  requires manual editing of objects. This will be significantly more difficult with the
+  backend serialized in the wire format.
+
+##### Write custom storage objects for the backend
+
+Rather than use gRPC's wire format or serialized JSON format, we should consider writing
+a storage object format. This has a number of benefits:
+
+- The backend continues to be easy to read and modify.
+- We decouple our protobuf definitions from our storage representation, which makes changes
+  to the frontend less impactful.
+- This can be done without making any changes to the proto definitions.
+- It gives us more control over storage representation.
+- It makes migration easier, as we can easily maintain multiple versions of an object within
+  the Teleport codebase.
+
+The downsides:
+
+- It adds to the boilerplate necessary for an object.
 
 ##### Write custom marshalers for objects
 
 In order to keep our user facing representations of the objects, we will need to write
 custom JSON and YAML marshalers for our objects. This may be time intensive but not overall
 a significant amount of work.
 
+#### Embedding types (removing `gogoproto.embed`)
```

Comment:
Changes like this will cause Go-level incompatible changes in api/. 

I think we should discuss whether these can be avoided and, if not (or not worth it), what kind of "migration" path we are thinking of implementing. (Ie, how we'll deal with customer breakages and communication, potential e/ breakages, etc.)

Reply:
Hrm. I'm not sure we can avoid breakages without keeping gogo intact, or at least a segment of Teleport that still maintains has it. Even getting rid of `nullable` will cause a Go level breakage here. Maybe we could have an `api-legacy` that we maintain for 3 versions and then remove?

I don't think we can avoid e breakages without keeping the object interfaces intact, or without doing some deep wizardry on the backend service interfaces, which sounds hideous.

Customer breakages should definitely be well communicated, but I'm not sure of the best mechanism there.

Reply:
Communication and a plan seem good enough in this case. It's worth noting that this could cause pains for a few versions while everything is moved to new formats.

Reply:
I don't think we currently care much about code-level incompatibility in the API module, fwiw.

Reply:
I think we caused breakages for customers a few times because of API changes. @zmb3 for feedback on how much we should worry about this.